io/io/src/TFileMergeServer.cxx

```diff
@@ -0,0 +1,348 @@
+{
+   if (destination == 0 || source == 0) return;
+   TIter nextkey(source->GetListOfKeys());
+
+   TKey *key;
+
+   while ((key = (TKey *)nextkey())) {
+      TClass *cl = TClass::GetClass(key->GetClassName());
+      if (cl->InheritsFrom(TDirectory::Class())) {
+         TDirectory *source_subdir = (TDirectory *)source->GetList()->FindObject(key->GetName());
+         if (!source_subdir) {
+            source_subdir = (TDirectory *)key->ReadObj();
+         }
+         TDirectory *destination_subdir = destination->GetDirectory(key->GetName());
+         if (!destination_subdir) {
+            destination_subdir = destination->mkdir(key->GetName());
+         }
+         R__MigrateKey(destination, source);
+      } else {
+         TKey *oldkey = destination->GetKey(key->GetName());
+         if (oldkey) {
+            oldkey->Delete();
+            delete oldkey;
+         }
+         // a priori the files are from the same client
+         TKey *newkey = new TKey(destination, *key, 0 /* pidoffset */);
+         destination->GetFile()->SumBuffer(newkey->GetObjlen());
+         newkey->WriteFile(0);
+         if (destination->GetFile()->TestBit(TFile::kWriteError)) {
+            return;
+         }
+      }
+   }
+   destination->SaveSelf();
+}
+
+struct ClientInfo {
+   TFile *fFile; // This object does *not* own the file, it's owned by the owner of ClientInfo
+   TString fLocalName;
+   UInt_t fContactsCount;
+   TTimeStamp fLastContact;
+   Double_t fTimeSincePrevContact;
+
+   ClientInfo() : fFile(0), fLocalName(), fContactsCount(0), fTimeSincePrevContact(0) {}
+
+   ClientInfo(const char *filename, UInt_t clientId)
+      : fFile(0), fContactsCount(0), fTimeSincePrevContact(0)
+   {
+      fLocalName.Form("%s-%d-%d", filename, clientId, gSystem->GetPid());
+   }
+
+   void Set(TFile *file)
+   {
+      // Register the new file as coming from this client.
+      if (file != fFile) {
+         // We need to keep any of the keys from the previous file that
+         // are not in the new file.
+         if (fFile) {
+            R__MigrateKey(fFile, file);
+            // delete the previous memory file (if any)
+            delete file;
+         } else {
+            fFile = file;
+         }
+      }
+      TTimeStamp now;
+      fTimeSincePrevContact = now.AsDouble() - fLastContact.AsDouble();
+      fLastContact = now;
+      ++fContactsCount;
+   }
+};
+
+struct ThreadFileMerger : public TObject {
+   TString fFilename;
+   TBits fClientsContact;
+   UInt_t fNClientsContact;
+   std::vector<ClientInfo> fClients;
+   TTimeStamp fLastMerge;
+   TFileMerger fMerger;
+
+   ThreadFileMerger(const char *filename, Bool_t writeCache = false)
+      : fFilename(filename), fNClientsContact(0), fMerger(false, true)
+   {
+      fMerger.SetPrintLevel(0);
+      fMerger.OutputFile(filename, "RECREATE");
+      if (writeCache)
+         new TFileCacheWrite(fMerger.GetOutputFile(), 32 * 1024 * 1024);
```

Comment:
This is a widely known ROOT behaviour. Some objects attach themselves to the gDirectory. If a change shall take place, it will be for ROOT7.